### PR TITLE
Rust: Attribute-style procedural macro to generate EVMC FFI code with user-defined data

### LINKED ISSUE #168: Create Rust binding
See also https://github.com/ETCDEVTeam/sputnikvm-ffi/issues/12

### PR DESCRIPTION
**Pull request by jakelang**
Closes #168.
Closes #223.

### CODE REVIEW

File: bindings/rust/evmc-declare-tests/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+use evmc_vm::EvmcVm;
+use evmc_vm::ExecutionResult;
+use evmc_vm::ExecutionContext;
+#[macro_use]
+use evmc_declare::evmc_declare_vm;
+
+#[evmc_declare_vm(FooVm = "ewasm")]
```
Comment (axic) [inline bindings/rust/evmc-declare-tests/src/lib.rs]: Is it possible to merge `evmc-declare-tests` into a test section in `evmc-declare`?

File: bindings/rust/evmc-vm/src/lib.rs [hunk 2]
```diff
@@ -9,6 +9,11 @@ pub use evmc_sys as ffi;
 // TODO: Add convenient helpers for evmc_execute
 // TODO: Add a derive macro here for creating evmc_create
```
Comment (axic) [inline bindings/rust/evmc-vm/src/lib.rs]: Both of these should be removed.

File: bindings/rust/evmc-declare-tests/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+impl EvmcVm for FooVM {
+    fn init() -> Self {
+        FooVM { 
+            a: 105023,
+        }
+    }
+
+    fn execute(&self, code: &[u8], context: &ExecutionContext) -> ExecutionResult {
```
Comment (axic) [inline bindings/rust/evmc-declare-tests/src/lib.rs]: As a next step perhaps we could use the `Result` pattern, where `Error` is translated into a failure code in `ExecutionResult`. But I wouldn't touch that in this PR.

File: bindings/rust/evmc-declare/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+                (name, capabilities)
+            } else {
+                panic!("Argument passed to evmc_declare_vm is not a name-value item.")
+            }
+        }
+        _ => panic!("Argument passed to evmc_declare_vm is of incorrect type. Expected a name-value item.")
+    };
+    
+    // Get the VM version from the crate version.
+    let vm_version_string = env!("CARGO_PKG_VERSION").to_string();
```
Comment (axic) [inline bindings/rust/evmc-declare/src/lib.rs]: While I like this we need to document it somehow.

File: bindings/rust/evmc-declare/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+    // Extract the name and capabilities from the attribute's key-value item.
+    let (vm_name_stylized, vm_capabilities) = match &meta.nested[0] {
+        NestedMeta::Meta(m) => {
+            // Try to form a string from the identifier if a meta-item was supplied.
+            if let Meta::NameValue(pair) = m {
+                let name = pair.ident.to_string();
+                let capabilities = match pair.lit {
+                    Lit::Str(ref s) => match s.value().as_str() {
+                        "evm1" => 0x1u32,
+                        "ewasm" => 0x1u32 << 1u32,
```
Comment (axic) [inline bindings/rust/evmc-declare/src/lib.rs]: It seems it only allows a single capability currently?

File: bindings/rust/evmc-declare/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+}
+
+/// Generates a definition and impl for a struct which contains the EVMC instance needed by FFI,
+/// and the user-defined VM.
+// TODO: Move this struct and impl into evmc_vm.
+fn build_vm_container() -> proc_macro2::TokenStream {
+    quote! {
+        struct __EvmcContainer<T: ::evmc_vm::EvmcVm + Sized> {
+            instance: ::evmc_sys::evmc_instance,
+            vm: T,
```
Comment (axic) [inline bindings/rust/evmc-declare/src/lib.rs]: So we're still doing the same trick here :wink:

File: bindings/rust/evmc-declare/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+
+    quote! {
+        static #static_name_ident: &'static str = #stylized_name_literal;
+        static #static_version_ident: &'static str = #version_literal;
+    }
+}
+
+/// Generates a definition and impl for a struct which contains the EVMC instance needed by FFI,
+/// and the user-defined VM.
+// TODO: Move this struct and impl into evmc_vm.
```
Comment (axic) [inline bindings/rust/evmc-declare/src/lib.rs]: How would you do that?

**Comment by jakelang** [comment]
It is not macro-parameterized, so we can just stick the entire thing in
evmc-vm's lib.rs.

File: bindings/rust/evmc-declare/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+    }
+}
+
+/// Takes a capabilities flag and builds the evmc_get_capabilities callback.
+fn build_capabilities_fn(name_lowercase: &String, capabilities: u32) -> proc_macro2::TokenStream {
+    // Could avoid using a special name and just use get_capabilities.
+    let concatenated = format!("{}_get_capabilities", name_lowercase);
+    let capabilities_fn_ident = Ident::new(&concatenated, name_lowercase.span());
+    let capabilities_literal =
+        LitInt::new(capabilities as u64, IntSuffix::U32, capabilities.span());
```
Comment (axic) [inline bindings/rust/evmc-declare/src/lib.rs]: Why `u64` when it says `IntSuffix::U32`? Looks weird.

File: bindings/rust/evmc-declare/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+    quote! {
+        extern "C" fn #fn_ident(instance: *mut ::evmc_sys::evmc_instance) {
+            Box::new(__EvmcContainer::from_ffi_pointer::<#type_ident>(instance));
+        }
+    }
+}
+
+/// Takes a capabilities flag and builds the evmc_get_capabilities callback.
+fn build_capabilities_fn(name_lowercase: &String, capabilities: u32) -> proc_macro2::TokenStream {
+    // Could avoid using a special name and just use get_capabilities.
```
Comment (axic) [inline bindings/rust/evmc-declare/src/lib.rs]: Yep, the special name is only required for `create` as that is the public symbol, everything else should be scoped for the file.

The only benefit we get with prefixes is that multiple VMs could be declared in the same file, but not sure we want that.

**Comment by jakelang** [comment]
I switched back in order to get this done more quickly. To do the field
manipulation (i had this done in earlier commits), we also need to be able
to properly initialize arbitrary fields in evmc_create_vm

File: bindings/rust/evmc-declare/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+
+            let code_ref: &[u8] = unsafe {
+                ::std::slice::from_raw_parts(code, code_size);
+            }
+
+            let container = __EvmcContainer::from_ffi_pointer::<#type_name_ident>(instance);
+
+            let result = container.execute(code_ref, &execution_context);
+
+            container.into_ffi_pointer();
```
Comment (axic) [inline bindings/rust/evmc-declare/src/lib.rs]: Not sure why shuffling this one struct around is needed? Also `from_ffi_pointer` seems to create a clone first?

File: bindings/rust/evmc-declare/src/lib.rs (hunk 5 above)
Comment (axic) [reply bindings/rust/evmc-declare/src/lib.rs]: Also this should use values from `evmc::ffi::EVMC_CAPABILITIES` and not hardcoded ones.

**Comment by axic** [comment]
Also we need to change the examplevm to use this in order to test it.

File: bindings/rust/evmc-declare/src/lib.rs (hunk 8 above)
Comment (jakelang) [reply bindings/rust/evmc-declare/src/lib.rs]: because the constructor takes a u64, independent of the actual width of the literal we are trying to construct. It will, nonetheless, produce a U32 literal.

File: bindings/rust/evmc-declare/src/lib.rs (hunk 9 above)
Comment (jakelang) [reply bindings/rust/evmc-declare/src/lib.rs]: Done.

File: bindings/rust/evmc-declare/src/lib.rs (hunk 7 above)
Comment (jakelang) [reply bindings/rust/evmc-declare/src/lib.rs]: Done.

File: bindings/rust/evmc-declare/src/lib.rs (hunk 5 above)
Comment (jakelang) [reply bindings/rust/evmc-declare/src/lib.rs]: Allows multiple capabilities.

Comment (jakelang) [reply bindings/rust/evmc-declare/src/lib.rs]: also, using the enum and playing with the discriminants doesn't make this cleaner IMO.

File: bindings/rust/evmc-declare/src/lib.rs (hunk 4 above)
Comment (jakelang) [reply bindings/rust/evmc-declare/src/lib.rs]: Yeah, not sure where to.

File: bindings/rust/evmc-declare-tests/src/lib.rs (hunk 3 above)
Comment (jakelang) [reply bindings/rust/evmc-declare-tests/src/lib.rs]: Yeah, later

File: bindings/rust/evmc-vm/src/lib.rs (hunk 2 above)
Comment (jakelang) [reply bindings/rust/evmc-vm/src/lib.rs]: Done

File: bindings/rust/evmc-declare-tests/src/lib.rs (hunk 1 above)
Comment (jakelang) [reply bindings/rust/evmc-declare-tests/src/lib.rs]: No, you cannot test a procedural macro in the same crate :(
`proc_macro2` does have some facilities for this but I prefer to test via code expansion, etc.

File: bindings/rust/evmc-declare/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+
+            let code_ref: &[u8] = unsafe {
+                ::std::slice::from_raw_parts(code, code_size)
+            };
+
+            let container = unsafe {
+                ::evmc_vm::EvmcContainer::<#type_name_ident>::from_ffi_pointer(instance)
+            };
+
+            let result = container.execute(code_ref, &execution_context);
```
Comment (axic) [inline bindings/rust/evmc-declare/src/lib.rs]: Should wrap this in panic unwinding and return `EVMC_INTERNAL_ERROR`:
```
panic::catch_unwind(|| {
    container.execute(..)
});
```

Comment (axic) [reply bindings/rust/evmc-declare/src/lib.rs]: Well actually we should do that for the entire `execute` function except the `assert!`s on top.

Comment (axic) [reply bindings/rust/evmc-declare/src/lib.rs]: And potentially for panics caused but this macro should be ensured that result in abort and not unwinding.

File: bindings/rust/evmc-declare-tests/src/lib.rs (hunk 1 above)
Comment (axic) [reply bindings/rust/evmc-declare-tests/src/lib.rs]: Ok, lets keep the extra crates then.

File: bindings/rust/evmc-declare-tests/src/lib.rs (hunk 3 above)
Comment (axic) [reply bindings/rust/evmc-declare-tests/src/lib.rs]: Created issue #309 to track this.

File: bindings/rust/evmc-declare/src/lib.rs (hunk 10 above)
Comment (axic) [reply bindings/rust/evmc-declare/src/lib.rs]: Was there an answer to this or the code is gone (haven't checked)?

**Comment by axic** [comment]
>  NOTE: Removed the evmc_sys as ffi alias. May reintroduce later if I make a workaround.

Why? Can it be done in a separate PR then?

**Comment by axic** [comment]
Have you tested that the bumpversion config works?

File: bindings/rust/evmc-declare/Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,17 @@
+[package]
+name = "evmc-declare"
```
Comment (axic) [inline bindings/rust/evmc-declare/Cargo.toml]: Should this be called declare or derive?

File: bindings/rust/evmc-declare/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,310 @@
+#![recursion_limit = "128"]
```
Comment (axic) [inline bindings/rust/evmc-declare/src/lib.rs]: Why is this needed?

File: bindings/rust/evmc-declare/src/lib.rs (hunk 11 above)
Comment (axic) [reply bindings/rust/evmc-declare/src/lib.rs]: Created issue #310 to track this.

File: bindings/rust/evmc-vm/src/lib.rs [hunk 14]
```diff
@@ -4,30 +4,35 @@
  */
 
 pub extern crate evmc_sys;
-pub use evmc_sys as ffi;
```
Comment (axic) [inline bindings/rust/evmc-vm/src/lib.rs]: This seems to be a trend among `-sys` crates, that's the reason I did it.

File: examples/example-rust-vm/src/lib.rs [hunk 15]
```diff
@@ -2,62 +2,25 @@ extern crate evmc_vm;
...
-    code_size: usize,
-) -> ffi::evmc_result {
-    let execution_ctx = unsafe {
-        ExecutionContext::new(
-            msg.as_ref().expect("tester passed nullptr as message"),
-            context.as_mut().expect("tester passed nullptr as context"),
-        )
-    };
-    let is_create = execution_ctx.get_message().kind == ffi::evmc_call_kind::EVMC_CREATE;
+#[evmc_declare_vm("ExampleRustVM", "evm")]
```
Comment (axic) [inline examples/example-rust-vm/src/lib.rs]: Is `evm` the version here? If so, we should use a usual version string because otherwise it is confusing, e.g. "0.1.0".

Comment (jakelang) [reply examples/example-rust-vm/src/lib.rs]: that's the capabilities set. It is a `", "` -separated list of capabilities, which at the moment includes `"ewasm"` and `"ewasm"`. I need to document somewhere visible.

File: bindings/rust/evmc-vm/src/lib.rs (hunk 14 above)
Comment (jakelang) [reply bindings/rust/evmc-vm/src/lib.rs]: I will re-introduce it in a bit. I just needed to get rid of it for the moment to organize the visibility layout for the procedural macro.

File: bindings/rust/evmc-declare/src/lib.rs (hunk 13 above)
Comment (jakelang) [reply bindings/rust/evmc-declare/src/lib.rs]: The default recursion limit is `64`, which at the moment has not proven viable for parsing some structures in this macro.

File: bindings/rust/evmc-declare/Cargo.toml (hunk 12 above)
Comment (jakelang) [reply bindings/rust/evmc-declare/Cargo.toml]: This isn't a derive macro. It is an attribute macro. `derive`-mode macros auto-impl a trait, while this macro does not. It generates FFI code for calling the EVMC VM to which it is applied.

**Comment by jakelang** [comment]
> 
> 
> Have you tested that the bumpversion config works?

How do I test this?

File: bindings/rust/evmc-declare/src/lib.rs (hunk 11 above)
Comment (jakelang) [reply bindings/rust/evmc-declare/src/lib.rs]: I'd be cautious about trying to introduce unwinding infrastructure this early. Maybe in a separate PR.

File: examples/example-rust-vm/src/lib.rs (hunk 15 above)
Comment (jakelang) [reply examples/example-rust-vm/src/lib.rs]: The version is the crate version. I may add an option to specify a crate-independent version, as I have figured out how to make this possible.

File: bindings/rust/evmc-vm/src/lib.rs (hunk 14 above)
Comment (axic) [reply bindings/rust/evmc-vm/src/lib.rs]: I don't fully understand why it needs to be removed now and be readded later introducing a lot of changes across the way?

File: examples/example-rust-vm/src/lib.rs (hunk 15 above)
Comment (axic) [reply examples/example-rust-vm/src/lib.rs]: Well these should be rustdoc documented in the `evmc-declare` crate as well as with a comment in this example file.

File: bindings/rust/evmc-declare/src/lib.rs (hunk 13 above)
Comment (axic) [reply bindings/rust/evmc-declare/src/lib.rs]: Can you leave this comment above the setting?

File: bindings/rust/evmc-declare/Cargo.toml (hunk 12 above)
Comment (axic) [reply bindings/rust/evmc-declare/Cargo.toml]: Wasn't this a derive macro initially?

File: bindings/rust/evmc-declare/src/lib.rs (hunk 10 above)
Comment (jakelang) [reply bindings/rust/evmc-declare/src/lib.rs]: It is not cloned anymore.

File: bindings/rust/evmc-declare-tests/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,20 @@
+use evmc_vm::EvmcVm;
+use evmc_vm::ExecutionContext;
+use evmc_vm::ExecutionResult;
+#[macro_use]
+use evmc_declare::evmc_declare_vm;
+
+#[evmc_declare_vm("Foo VM", "ewasm")]
```
Comment (axic) [inline bindings/rust/evmc-declare-tests/src/lib.rs]: Since this is a comprehensive test, I'd say we need to VMs declared here:
1) supports single capability and takes version from crate
2) supports multiple capabilities and takes version from declare macro

File: bindings/rust/evmc-declare-tests/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+    a: i32,
+}
+
+impl EvmcVm for FooVM {
+    fn init() -> Self {
+        FooVM { a: 105023 }
+    }
+
+    fn execute(&self, code: &[u8], context: &ExecutionContext) -> ExecutionResult {
+        ExecutionResult::new(evmc_sys::evmc_status_code::EVMC_SUCCESS, 235117, None)
```
Comment (axic) [inline bindings/rust/evmc-declare-tests/src/lib.rs]: Could just use `ExecutionResult::success()` here.

File: bindings/rust/evmc-declare/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,383 @@
+//! evmc-declare is an attribute-style procedural macro to be used for automatic generation of FFI
+//! code for the EVM-C API with minimal boilerplate.
```
Comment (axic) [inline bindings/rust/evmc-declare/src/lib.rs]: EVMC, not EVM-C

File: bindings/rust/evmc-declare/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,383 @@
...
+    quote! {
+        static #static_name_ident: &'static str = #stylized_name_literal;
+        static #static_version_ident: &'static str = #version_tokens;
+    }
+}
+
+/// Takes a capabilities flag and builds the evmc_get_capabilities callback.
+fn build_capabilities_fn(capabilities: u32) -> proc_macro2::TokenStream {
+    let capabilities_literal =
+        LitInt::new(capabilities as u64, IntSuffix::U32, capabilities.span());
```
Comment (axic) [inline bindings/rust/evmc-declare/src/lib.rs]: Clippy gives this:
```
warning: casting u32 to u64 may become silently lossy if types change
   --> bindings/rust/evmc-declare/src/lib.rs:281:21
    |
281 |         LitInt::new(capabilities as u64, IntSuffix::U32, capabilities.span());
    |                     ^^^^^^^^^^^^^^^^^^^ help: try: `u64::from(capabilities)`
```

Comment (jakelang) [reply bindings/rust/evmc-declare/src/lib.rs]: This is expected. This is a problem with `syn` because it takes a u64 for any integer literal type. In our case, there should be no issue.

Comment (axic) [reply bindings/rust/evmc-declare/src/lib.rs]: Is `u64::from` nicer?
